### PR TITLE
pad the end batch

### PR DESCRIPTION
**Pull request by yajiedesign**

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -850,6 +853,18 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
                 batch_logs = {}
                 batch_logs['batch'] = batch_index
                 batch_logs['size'] = len(batch_ids)
+                if K.backend() == 'mxnet':
```
Comment (piiswrong) [inline keras/engine/training.py]: dont pad in fit_loop. do it in train_function by calling a utility function, so you don't need to do it again in test

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -1880,7 +1898,11 @@ def conv2d(x, kernel, strides=(1, 1), border_mode='valid',
     # Returns
         A tensor, result of 2D convolution.
     """
-    raise NotImplementedError
+    shape = kernel.shape
```
Comment (howard0su) [inline keras/backend/mxnet_backend.py]: you need handle dim_ordering. the following code can handle 3d tensor.


def _layout_from(dim_ordering):
    if dim_ordering == 'default':
        dim_ordering = image_dim_ordering()
    if dim_ordering == 'th':
        return 'NHWC'
    elif dim_ordering == 'tf':
        return 'NCHW'
    else:
        raise ValueError('Unknown dim_ordering ' + str(dim_ordering))

File: keras/engine/training.py [hunk 3]
```diff
@@ -230,9 +231,9 @@ def check_loss_and_target_compatibility(targets, losses, output_shapes):
                         'A target array with shape ' + str(y.shape) +
                         ' was passed for an output of shape ' + str(shape) +
                         ' while using as loss `' + loss.__name__ + '`. '
-                        'This loss expects '
-                        'targets to have the same shape '
-                        'as the output.')
+                                                                   'This loss expects '
```
Comment (howard0su) [inline keras/engine/training.py]: please revert these formatting.

File: keras/engine/training.py [hunk 4]
```diff
@@ -337,6 +339,7 @@ def weighted(y_true, y_pred, weights, mask=None):
             score_array *= weights
             score_array /= K.mean(K.cast(K.not_equal(weights, 0), K.floatx()))
         return K.mean(score_array)
+
```
Comment (howard0su) [inline keras/engine/training.py]: this is indeed correct change. but still revert it to avoid headache of merge

File: keras/engine/training.py [hunk 5]
```diff
@@ -1785,28 +1789,47 @@ def compile(self, optimizer, loss, metrics=None, loss_weights=None,
...
             self._pred_mod = K.mx.mod.Module(symbol=self._symbol, data_names=self._data_names,
                                              label_names=None, context=self._context)
             self._weights_dirty = False
+            self.batch_size = None
 
             def train_function(inputs):
+                size = 0
+                if self.batch_size is not None:
+                    size = self.batch_size - inputs[0].shape[0]
+                if size > 0:
```
Comment (howard0su) [inline keras/engine/training.py]: please abstract as a function and this logic need happen in predict_function

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 from __future__ import print_function
 import mxnet as mx
+from mxnet.symbol_doc import SymbolDoc
```
Comment (howard0su) [inline keras/backend/mxnet_backend.py]: what is this for?

File: keras/backend/mxnet_backend.py [hunk 7]
```diff
@@ -1880,7 +1897,10 @@ def conv2d(x, kernel, strides=(1, 1), border_mode='valid',
     # Returns
         A tensor, result of 2D convolution.
     """
-    raise NotImplementedError
+    layout_kernel = _layout_kernel(dim_ordering, kernel.shape)
+    s = mx.sym.Convolution(data=x.symbol, name=kernel.name, kernel=layout_kernel, stride=strides,
```
Comment (howard0su) [inline keras/backend/mxnet_backend.py]: you need pass layout as well.

File: keras/engine/training.py [hunk 8]
```diff
@@ -448,8 +450,8 @@ def data_generator_task():
 
     return q, _stop, generator_threads
 
-class Model(Container):
 
+class Model(Container):
```
Comment (howard0su) [inline keras/engine/training.py]: can u reduce this diff?

**Pull request merged**